[Daniel Tan Ren Jie] iP

cookiedan42 opened the pull request:
# DAN 9000
> I'm sorry Dave, I'm afraid I can't do that
> - [HAL 9000](https://www.youtube.com/watch?v=ARJ8cAGm6JE) (~~hopefully not related~~)  

On a very important mission that cannot be jeopardized?
Let **DAN 9000** help you keep track of the essential tasks!  

Steps
1. Download the latest release
1. Double click it to run `Main`
1. ???
1. Profit  

- ok?
   * ok.  
   * ok.

```java
public class DukeException extends Exception {
    public DukeException(String message) {
        super(message);
    }
}

```
TODO list
- [x] Build task list
- [ ] Ensure user safety :safety_vest:

Commit:
Add Gradle support

cookiedan42 committed:
add Greet functionality

cookiedan42 committed:
add Exit functionality

cookiedan42 committed:
Add Echo functionality

cookiedan42 committed:
add Add functionality

cookiedan42 committed:
add List functionality

cookiedan42 committed:
add Mark as Done functionality

cookiedan42 committed:
refactor printing to take String instead of String[]

cookiedan42 committed:
fix TaskArrayList markDone range

cookiedan42 committed:
add Todo functionality

cookiedan42 committed:
added Deadline functionality

cookiedan42 committed:
add Event functionality

cookiedan42 committed:
refactor cases

cookiedan42 committed:
change logo to only use characters that can be written to .txt

cookiedan42 committed:
populate text ui testing

cookiedan42 committed:
refactor commands into switch case structure

cookiedan42 committed:
add DukeException

cookiedan42 committed:
migrate done to use DukeException

cookiedan42 committed:
migrate todo to use DukeException

cookiedan42 committed:
migrate deadline to use DukeException

cookiedan42 committed:
migrate event to use DukeException

cookiedan42 committed:
move add task logic into TaskArrayList

cookiedan42 committed:
implement delete functionality

cookiedan42 committed:
Add serialise method to Tasks

cookiedan42 committed:
Add save and load functionality

Save file is structured as json-like text file.
header is Task:<task type>
followed by each argument on a new line as Arg:<value>
tasks are separated by an empty line

eg :
Task:todo
\tName:1100
\tDone:false

Task:deadline
\tName:110022
\tDone:false
\tBy: now

cookiedan42 committed:
Change by into a countdown for deadline tasks when data is a datetime

cookiedan42 committed:
Merge branch 'branch-Level-7' into master

cookiedan42 committed:
Merge branch 'branch-Level-8' into master

cookiedan42 committed:
Remove leftover print statement

cookiedan42 committed:
Rename Parser to SaveParser

Deconflict names with A-MoreOOP requirements

cookiedan42 committed:
Move simple printing into UI class

Move formatted text string and static printing of logo and welcome text

cookiedan42 committed:
Move command selection from Duke into Parser

cookiedan42 committed:
Move command parsing logic from task into Parser

cookiedan42 committed:
Move delete command logic into Parser from TaskListArray

cookiedan42 committed:
Move exiting control back to Duke from Parser

cookiedan42 committed:
Move mark done command parsing into Parser

cookiedan42 committed:
Reorganize code in packages

cookiedan42 committed:
Break up Parser logic into methods

cookiedan42 committed:
Break up SaveParser logic into separate methods

cookiedan42 committed:
Clean up autorefactor error

cookiedan42 committed:
Add find task functionality

cookiedan42 committed:
Reformat code to follow coding style

cookiedan42 committed:
Add javadocs

cookiedan42 committed:
Merge branch 'branch-A-JavaDoc' into master

cookiedan42 committed:
Merge branch 'branch-A-CodingStandard' into master

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskArrayList.java
#	src/main/java/duke/UI.java
#	src/main/java/duke/exceptions/DukeException.java
#	src/main/java/duke/storage/SaveParser.java
#	src/main/java/duke/tasks/Deadline.java
#	src/main/java/duke/tasks/Event.java
#	src/main/java/duke/tasks/Task.java
#	src/main/java/duke/tasks/Todo.java

cookiedan42 committed:
Merge branch 'branch-Level-9' into master

# Conflicts:
#	src/main/java/duke/TaskArrayList.java

cookiedan42 committed:
Add minimal Tests

cookiedan42 committed:
Merge branch 'add-gradle-support' into A-Gradle

cookiedan42 committed:
Modify build.gradle for project name

cookiedan42 committed:
Merge branch 'branch-A-Gradle' into master

cookiedan42 committed:
Add basic GUI functionality bootstrapped off CLI responses

atyhamos commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,51 @@
+ * Main class for Duke.
+ */
+public class Duke {
+    static String border = "--------------------------------------------------";
```
You could make this into a constant 😄

eezj35 replied:
I agree with atyhamos's comment!

atyhamos commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,51 @@
+            taskList = new TaskArrayList();
+            ui.showLoadingError();
+        }
+        ui.displayLogo();
+        ui.displayWelcome();
```
I like that these are easy to read and understand 😄

atyhamos commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,154 @@
+        Task toAdd;
+
+        // switch case to split by command
+        switch (cmd_args[0]) {
+        case ("bye"):
+            this.runBye(cmd_args);
+            return true;
+
+        case ("list"):
+            runList(cmd_args);
+            break;
+
+        case ("done"):
+            runDone(cmd_args);
+            break;
+
+        case ("delete"):
+            runDelete(cmd_args);
+            break;
+
+        case ("find"):
+            runFind(cmd_args);
+            break;
+
+        case ("todo"):
+            runTodo(cmd_args);
+            break;
+
+        case ("deadline"):
+            runDeadline(cmd_args);
+            break;
+
+        case ("event"):
+            runEvent(cmd_args);
+            break;
+
+        default:
+            throw new DukeException("Unrecognised command");
+        }
```
I like this too! Simple and easy to read.

eezj35 replied:
Agree with atyhamos, and I like the spaces between each case block, it is very helpful 👍

atyhamos commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    private void runFind(String[] cmd_args) throws DukeException {
+        if (cmd_args.length != 2) {
+            throw new DukeException(TaskArrayList.FIND_USAGE_TEXT);
+        }
+        ui.display(taskList.find(cmd_args[1]));
+    }
+
+    private void runTodo(String[] cmd_args) throws DukeException {
+        if (cmd_args.length != 2) {
+            throw new DukeException(Todo.USAGE_TEXT);
+        }
+        Task toAdd = new Todo(cmd_args[1]);
+        ui.display(taskList.addTask(toAdd));
+    }
+
+    private void runDeadline(String[] cmd_args) throws DukeException {
+        if (cmd_args.length != 2) {
+            throw new DukeException(Deadline.USAGE_TEXT);
+        }
+        String[] name_by = cmd_args[1].split("/by", 2);
+        if (name_by.length != 2) {
+            throw new DukeException(Deadline.USAGE_TEXT);
+        }
+        Task toAdd = new Deadline(name_by[0], name_by[1]);
+        ui.display(taskList.addTask(toAdd));
+    }
+
+    private void runEvent(String[] cmd_args) throws DukeException {
+        if (cmd_args.length != 2) {
+            throw new DukeException(Event.USAGE_TEXT);
+        }
+        String[] name_at = cmd_args[1].split("/at", 2);
+        if (name_at.length != 2) {
+            throw new DukeException(Event.USAGE_TEXT);
+        }
+        Task toAdd = new Event(name_at[0], name_at[1]);
+        ui.display(taskList.addTask(toAdd));
+    }
```
Perhaps to avoid filling this Parser class with too many methods outside of its job as a "Parser", you could implement these methods in separate Command classes under a parent "Command" class?

atyhamos commented on src/main/java/duke/TaskArrayList.java:
```diff
@@ -0,0 +1,106 @@
+ * Array List of Task objects.
+ */
+public class TaskArrayList extends ArrayList<Task> {
+    public final static String DELETE_USAGE_TEXT = "Usage: delete <integer task number>";
+    public final static String DONE_USAGE_TEXT = "Usage: done <integer task number>";
+    public final static String FIND_USAGE_TEXT = "Usage: find <search String>";
```
I like that you included how to use the command as well

atyhamos commented on src/main/java/duke/TaskArrayList.java:
```diff
@@ -0,0 +1,106 @@
+        if (taskNo > this.size()) {
+            throw new DukeException(String.format("task %d not found", taskNo));
+        }
+        Task toDel = this.get(taskNo - 1);
```
Minor nitpick but it may be clearer and easier for other potential reviewers to understand if you named it toDelete.👍

atyhamos commented on src/main/java/duke/TaskArrayList.java:
```diff
@@ -0,0 +1,106 @@
+    /**
+     * Generates the message to print when tasklist length changes.
+     *
+     * @return You now have X task(s) in the list.
```
Your Javadoc comment doesn't match the output 😆

atyhamos commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,95 @@
+    /**
+     * Constructor for UI using default preprocessor.
+     */
+    UI() {
+        this(
+                (String source) -> {
+                    String border = "--------------------------------------------------";
+                    StringBuilder out = new StringBuilder();
+                    for (String line : source.split("\n")) {
+                        out.append("\t" + line + "\n");
+                    }
+                    return border + "\n"
+                            + out.toString()
+                            + border;
+                });
+    }
```
Perhaps you could amend this to make it less indented?

atyhamos commented on src/main/java/duke/storage/SaveParser.java:
```diff
@@ -0,0 +1,170 @@
+ * Parser class to create Task objects from string commands.
+ */
+public class SaveParser {
+    private static final DukeReadSaveException MISSING_ARG_EXCEPTION =
+            new DukeReadSaveException("missing argument in save");
```
I like this, quite smart to reuse a cached exception 👍

eezj35 replied:
Agree! Definitely will try this trick in my code in the future 👍

atyhamos commented on src/main/java/duke/storage/SaveParser.java:
```diff
@@ -0,0 +1,170 @@
+        String name = "";
+        boolean isDone = false;
+        String at = "";
+        boolean[] argsFound = new boolean[3];
```
Perhaps you could try to avoid using "Magic Numbers" like 3 here, as it is not obvious what significance it has.

atyhamos replied:
Or maybe include a comment so other reviewers would know why 3 was used 😄

eezj35 replied:
Yup I agree some comments would be good for others to more easily follow the code! 😄

atyhamos commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,68 @@
+     *
+     * @return added: task name.
+     */
+    public String addMsg() {
```
Perhaps change Msg to Message to make it easier to understand for those who may not recognise what Msg means

atyhamos reviewed (commented):
I like that your code is documented well with comments and that it follows the naming convention quite closely 👍 
There were parts where the code was nested heavily, and instances where the code looked confusing where you used magic numbers like 3 or 2,, but all in all I liked it 😄

eezj35 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,51 @@
+        Scanner userScanner = new Scanner(System.in);
+        Parser parser = new Parser(ui, taskList, storagePath);
+
+        boolean isExit = false;
```
Perhaps this variable name could be changed to canExit instead?

eezj35 commented on src/main/java/duke/TaskArrayList.java:
```diff
@@ -0,0 +1,106 @@
+     * Marks the given task as done.
+     *
+     * @param taskNo task number to mark as done.
+     * @return removal message String.
```
I believe the first letter of the descriptions for the param and return should be capitalized!

eezj35 commented on src/main/java/duke/TaskArrayList.java:
```diff
@@ -0,0 +1,106 @@
+    /**
+     * Enumerates members for printing
+     *
+     * @return String[] of "X. taskname".
```
I think to avoid misunderstanding, perhaps instead of using the term "String[]" to describe the return type, you could say something like"A String showing all the contents of the tasklist." Feel free to use or not use my suggestion though!

eezj35 commented on src/main/java/duke/TaskArrayList.java:
```diff
@@ -0,0 +1,106 @@
+     * enumerate but only include if task string contains search term
+     * use absolute task number so delete and done commands can reference this result
+     *
+     * @param searchTerm term to find in task names
```
I believe the "t" in "term" should be capitalized?

eezj35 commented on src/main/java/duke/TaskArrayList.java:
```diff
@@ -0,0 +1,106 @@
+     * use absolute task number so delete and done commands can reference this result
+     *
+     * @param searchTerm term to find in task names
+     * @return String[] of tasks of "X. taskname"
```
Same comment as the previous javadoc!

eezj35 commented on src/main/java/duke/TaskArrayList.java:
```diff
@@ -0,0 +1,106 @@
+    /**
+     * Marks a task as done.
+     *
+     * @param index task number to mark.
+     * @return mark done message String.
```
Same issue of capitalization of the first letter of the descriptions here!

eezj35 commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,95 @@
+     */
+    public void displayLogo() {
+        String logo =
+                "  _____          _   _    ___   ___   ___   ___  \n" +
+                " |  __ \\   /\\   | \\ | |  / _ \\ / _ \\ / _ \\ / _ \\ \n" +
+                " | |  | | /  \\  |  \\| | | (_) | | | | | | | | | |\n" +
+                " | |  | |/ /\\ \\ | . ` |  \\__, | | | | | | | | | |\n" +
+                " | |__| / ____ \\| |\\  |    / /| |_| | |_| | |_| |\n" +
+                " |_____/_/    \\_\\_| \\_|   /_/  \\___/ \\___/ \\___/ \n";
```
I like your logo, it's pretty cool and creative!

eezj35 commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,95 @@
+    /**
+     * Displays the provided string using preprocessor.
+     *
+     * @param content message String.
```
Noticed the same capitalization issue here! Perhaps you could check your other javadocs for this issue as well?

eezj35 commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @param name task name.
+     * @param done boolean state of task done.
+     * @param by   deadline for deadline task.
```
I think there are extra whitespaces between the "by" and the "deadline", perhaps you could delete the extra whitespaces?

eezj35 commented on src/test/java/duke/TaskArrayListTest.java:
```diff
@@ -0,0 +1,31 @@
+    @Test
+    void addTask_increaseSize() {
+        TaskArrayList taskList = new TaskArrayList();
+        assertEquals( 0,taskList.size());
```
I believe there should be a space after the comma, and also perhaps you could delete the extra whitespace in front of the "0" 😄

eezj35 commented on src/test/java/duke/UITest.java:
```diff
@@ -0,0 +1,66 @@
+package duke;
+
+
+
+
```
Perhaps you could delete the extra whitespaces here?

eezj35 commented on src/test/java/duke/UITest.java:
```diff
@@ -0,0 +1,66 @@
+    @Test
+    void displayLogo() {
+        String logo =
+                "  _____          _   _    ___   ___   ___   ___  \n" +
+                " |  __ \\   /\\   | \\ | |  / _ \\ / _ \\ / _ \\ / _ \\ \n" +
+                " | |  | | /  \\  |  \\| | | (_) | | | | | | | | | |\n" +
+                " | |  | |/ /\\ \\ | . ` |  \\__, | | | | | | | | | |\n" +
+                " | |__| / ____ \\| |\\  |    / /| |_| | |_| | |_| |\n" +
+                " |_____/_/    \\_\\_| \\_|   /_/  \\___/ \\___/ \\___/ \n\n";
```
Once again, love the logo 😄

eezj35 reviewed (commented):
Great job on the code overall! It was pretty clean and easy to read. I think the main issues with regards to the coding standards is just the javadocs standards, but other than that, I really liked your code!

cookiedan42 committed:
Add CSS formatting for dialog boxes

cookiedan42 committed:
Merge branch 'branch-Level-10' into master

cookiedan42 committed:
Add Checkstyle config files

cookiedan42 committed:
Tidy up code to follow checkstyle

cookiedan42 committed:
Merge branch 'branch-A-Checkstyle' into master

cookiedan42 committed:
Remove references to UI in standard GUI operation

cookiedan42 committed:
Add dedicated GUI launcher

cookiedan42 committed:
Add Assertions

Assertions are not included in the build

Assertions can help to show that program invariants always hold

Let's
* Enable assertions while running in gradle
* Write assertions for invariants we want to always hold

cookiedan42 committed:
Delete unused UI class and test

cookiedan42 committed:
Adhere to variable naming conventions

cookiedan42 committed:
Rewrite SaveParser for readability

SaveParser had repeated magic values low level code

This caused code duplication and difficulty reading

Let's
* Extract common patterns as methods
* Use static values for key values

cookiedan42 committed:
Tidy up Sorage Class

cookiedan42 committed:
Increase readability of runCases methods by adding descriptive condition names

cookiedan42 committed:
Improve Readability of TaskArrayList

cookiedan42 committed:
Merge pull request #1 from cookiedan42/branch-A-CodeQuality

Code Quality Task

cookiedan42 committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/duke/TaskArrayList.java
#	src/main/java/duke/storage/SaveParser.java

cookiedan42 committed:
Merge pull request #2 from cookiedan42/branch-A-Assertions

Branch a assertions

cookiedan42 committed:
Move Assertion to more appropriate location

cookiedan42 committed:
Add help Command

There is no inbuilt help for Duke

Let's
* Add help command to parser
* Add help Strings to command objects to serve as single source of truth
* Modify display window to accomodate the taller responses

cookiedan42 committed:
Add startup screen and tutorial

* Tutorial tasks appear as default when new save file generated
* Startup message to direct users to list and help commands

cookiedan42 committed:
Merge branch 'branch-C-Help' into master

cookiedan42 committed:
Add tests, Fix checkstyle

cookiedan42 committed:
Move parsing logic into separate command objects

Encapsulate all data for commands into individual classes to serve as single source of truth

cookiedan42 committed:
Delete profile images from chatbot

cookiedan42 committed:
Colorize Duke's response messages based on command validity

cookiedan42 committed:
Remove legacy entry point to duke

cookiedan42 committed:
Refactor Command classes to extract common methods and improve readability

cookiedan42 committed:
Clean up documentation

cookiedan42 committed:
Add initial User Guide version

cookiedan42 committed:
Tweak Header formatting

cookiedan42 committed:
Merge branch 'branch-UserGuide' into master

cookiedan42 committed:
Rename UI.png to Ui.png
